### PR TITLE
[SYCL][CUDA] libclc atomics update

### PR DESCRIPTION
**Pull request by Alexander-Johnston**
Updates the libclc spirv atomics used by the sycl cuda backend.

Includes
* Improvements to the bindings between libspirv and sycl for cuda
* Additional atomics
* Better testing in libclc for these atomics

### CODE REVIEW

File: sycl/include/CL/sycl/detail/spirv.hpp [hunk 1]
```diff
@@ -122,9 +122,12 @@ AtomicCompareExchange(multi_ptr<T, AddressSpace> MPtr,
...
   auto SPIRVScope = getScope(Scope);
-  auto *Ptr = MPtr.get();
-  return __spirv_AtomicCompareExchange(Ptr, SPIRVScope, SPIRVSuccess,
-                                       SPIRVFailure, Desired, Expected);
+  using I = detail::atomic_integer_arg_t<T>;
+  auto *Ptr = reinterpret_cast<typename multi_ptr<I, AddressSpace>::pointer_t>(
+      MPtr.get());
+
+  return (T)__spirv_AtomicCompareExchange(
+      Ptr, SPIRVScope, SPIRVSuccess, SPIRVFailure, (I)Desired, (I)Expected);
```
Comment (Pennycook) [inline sycl/include/CL/sycl/detail/spirv.hpp:130]: Are these changes necessary?  The `enable_if` should already limit `T` to integral types, and so it seems like `detail::atomic_integer_arg_t<T>` will only ever return `T` itself.

Same question for all the functions in this file.

Comment (StuartDAdams) [reply sycl/include/CL/sycl/detail/spirv.hpp:130]: This may require a bit of background. `clang/lib/Sema/SPIRVBuiltins.td` normalizes the builtin definitions an implementation can produce. Because it only exposes valid SPIR-V builtins, it protects the programmer. This all fits into the SPIR-V ecosystem - it represents a set of builtins that are usable in OpenCL and allows its implementation as a library. This means that projects other than SYCL can make use of the builtins, which will be good for SPIR-V and upstreaming. 

Because SYCL's atomic code is templated, it is possible that the user will attempt to use integer types that are not have corresponding SPIR-V atomic builtins. SPIR-V atomics do not provide implementations for `long long` or `unsigned long long` arguments, even though in practice they may be the same size as `long` or `unsigned long`. To work around this, we circumvent the type system so that `long long` and `unsigned long long` can be passed from SYCL to the spriv builtins. This PR introduces some machinery to pick an appropriately sized integer type and performs the casting as needed.

Comment (Pennycook) [reply sycl/include/CL/sycl/detail/spirv.hpp:130]: Ah, I see.  That makes sense -- thanks for the explanation.

Could we put a comment above the atomic functions (just one, before `AtomicCompareExchange`) that says something to this effect?  Something as simple as the below would do:
```
// C++ integer types are converted for SPIR-V built-ins
// e.g. unsigned long long => unsigned long
```

**Comment by elizabethandrews** [comment]
Not really familiar with this. @premanandrao @erichkeane can you take a look?

**Comment by bader** [comment]
@Alexander-Johnston, it looks like some functionality enabled by this patch is not supported by Intel OpenCL CPU device. Could you mark `regression/implicit_atomic_conversion.cpp` as expected to fail on CPU and ACC devices to unblock the merge, please?
We can investigate this issue later.

**Comment by Pennycook** [comment]
@bader: I don't think it's this patch's fault.  That regression test is for the functionality introduced by https://github.com/intel/llvm/pull/2044, and https://github.com/intel/llvm/pull/2127 is going to stop that functionality from being the default.

**Comment by bader** [comment]
@Pennycook: #2127 didn't help. The test is still failing.

**Comment by MrSidims** [comment]
> @Pennycook: #2127 didn't help. The test is still failing.

I'll take a look.

**Comment by MrSidims** [comment]
> > @Pennycook: #2127 didn't help. The test is still failing.
> 
> I'll take a look.

Purpose of the test is to check if compilation of a code with atomics that has implicit conversions for a fresh-new address spaces is working correct, so since the test in this PR is failing in runtime - https://github.com/intel/llvm/pull/2127 shall has no affect on it. Also I don't mind to disable runs of the on CPU/GPU. But the thing that worries me is that this change might break atomic conformance tests. Are we considering to deprecate atomics right at this point of time?

**Comment by bader** [comment]
> But the thing that worries me is that this change might break atomic conformance tests.

@MrSidims, is right. This change breaks `test_atomic` from Khronos SYCL-CTS.

**Comment by romanovvlad** [comment]
Please, fix tests and reopen the PR.
